### PR TITLE
Ray TPU Webhook Autoscaling Support

### PR DESCRIPTION
**Pull request by ryanaoleary**
This PR adds support for autoscaling RayClusters by generating `TPU_WORKER_HOSTNAMES` when intercepting each Pod, rather than when intercepting the RayCluster CR.

This PR has been tested as follows:

- [x] Unit Tests
 - [x] Manual Tests using [single-host](https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/ray-cluster.tpu-v4-singlehost.yaml), [multi-host](https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/ray-cluster.tpu-v4-multihost.yaml), and an [autoscaling RayCluster](https://github.com/ray-project/kuberay/blob/v1.1.1/ray-operator/config/samples/ray-cluster.autoscaler.yaml) with a TPU worker group added

### CODE REVIEW

File: ray-on-gke/tpu/kuberay-tpu-webhook/main.go [hunk 1]
```diff
@@ -134,18 +131,18 @@ func extractRayCluster(admissionReview *admissionv1.AdmissionReview) (*ray.RayCl
...
 }
 
-func genDNSHostnames(workerGroupSpec ray.WorkerGroupSpec, clusterName string, namespace string, replicaIndex int) (string, error) {
-	numHosts := workerGroupSpec.NumOfHosts
-	if numHosts == 0 {
-		return "", errors.New("workerGroupSpec NumOfHosts not set")
+func genDNSHostnames(numOfHosts int32, groupName string, clusterName string, namespace string, replicaIndex int) (string, error) {
+	if numOfHosts == 0 {
+		err := errors.New("workerGroupSpec NumOfHosts not set")
+		klog.ErrorS(err, "genDNSHostnames", "RayCluster", namespace+"/"+clusterName, "NumOfHosts", numOfHosts)
```
Comment (andrewsykim) [inline ray-on-gke/tpu/kuberay-tpu-webhook/main.go:137]: just return an error instead

File: ray-on-gke/tpu/kuberay-tpu-webhook/main.go [hunk 2]
```diff
@@ -134,18 +131,18 @@ func extractRayCluster(admissionReview *admissionv1.AdmissionReview) (*ray.RayCl
 	return &rayCluster, nil
 }
 
-func genDNSHostnames(workerGroupSpec ray.WorkerGroupSpec, clusterName string, namespace string, replicaIndex int) (string, error) {
-	numHosts := workerGroupSpec.NumOfHosts
-	if numHosts == 0 {
-		return "", errors.New("workerGroupSpec NumOfHosts not set")
+func genDNSHostnames(numOfHosts int32, groupName string, clusterName string, namespace string, replicaIndex int) (string, error) {
+	if numOfHosts == 0 {
```
Comment (andrewsykim) [inline ray-on-gke/tpu/kuberay-tpu-webhook/main.go:135]: numOfHosts = 0 is still valid with TPUs though right?

**Comment by andrewsykim** [comment]
> This PR adds support for autoscaling RayClusters by generating TPU_WORKER_HOSTNAMES when intercepting each Pod, rather than when intercepting the RayCluster CR.

Will existing pods need to have the `TPU_WORKER_HOSTNAMES` env var updated when new replicas are added or does it only apply for new pods added by the autoscaler?

File: ray-on-gke/tpu/kuberay-tpu-webhook/main.go (hunk 2 above)
Comment (ryanaoleary) [reply ray-on-gke/tpu/kuberay-tpu-webhook/main.go:135]: I don't think `numOfHosts = 0` is ever valid, `numOfHosts` defaults to 1 and if a user sets it to 0 then no workers would be created per replica.

**Comment by ryanaoleary** [comment]
> > This PR adds support for autoscaling RayClusters by generating TPU_WORKER_HOSTNAMES when intercepting each Pod, rather than when intercepting the RayCluster CR.
> 
> Will existing pods need to have the `TPU_WORKER_HOSTNAMES` env var updated when new replicas are added or does it only apply for new pods added by the autoscaler?

We don't need to update existing Pods. We assume that TPU PodSlices are scaled atomically so the `TPU_WORKER_HOSTNAMES` we add to each Pod include the DNS hostnames for all the Pods in the slice, even if they haven't been created yet.

File: ray-on-gke/tpu/kuberay-tpu-webhook/main.go (hunk 1 above)
Comment (ryanaoleary) [reply ray-on-gke/tpu/kuberay-tpu-webhook/main.go]: Done in [27ddd5d](https://github.com/GoogleCloudPlatform/ai-on-gke/pull/740/commits/27ddd5db325ab7a3856f5b8223e85d9fcb16295f).

**Review by andrewsykim — APPROVED** [review]

**Comment by andrewsykim** [comment]
(let's ensure https://github.com/GoogleCloudPlatform/ai-on-gke/pull/723 is also merged before cutting a new tag with the autoscaling changes)

**Pull request merged**